### PR TITLE
Eliminate storage views.

### LINKED ISSUE #46: Multiprocessing doesn't preserve data sharing of storage slices
``` python
x = torch.Storage(10)
y = x[1:-1]

#1
with open('file.t7', 'w+b') as f:
    torch.save([x, y], f)
    f.seek(0)
    a, b = torch.load(f)
# a and b no longer share the same storage

#2
q = multiprocessing.Queue()
q.put([x, y])
a, b = q.get()
# a and b no longer share the same data
```

### LINKED ISSUE #9447: Get rid of Storage View, IPC will not preserve storage structure
Storage views are currently used to implement CUDA IPC with the caching allocator. The basic problem is that you can only do IPC with a chunk of memory returned by cudaMalloc, but when you are using the caching CUDA allocator, lots of logically distinct allocations are together on the same cudaMalloc block, and you have no choice but to IPC the whole thing. A storage view was used to implement this situation, since you gave the entire CUDA caching allocator block a storage, and then you had the actual storages be views on this storage.

This is not a good reason to add something as complicated as storage views to PyTorch. So the new plan is this:

1. When you transfer a CUDA tensor across IPC, its storage/offsets change: instead of being the same storages they were before, you get the entire CUDA allocation storage.
2. The transferred tensors are non-resizable (this is true in practice today, since resizing would break sharing, but it is even more important after IPC, because a resize could cause you to clobber other allocations)

CC @apaszke, who I flew this plan by previously.

This should fix #5311

### PR DESCRIPTION
**Pull request by ezyang**
Storage views were previously used to implement CUDA IPC sharing,
but they weren't necessary.  The new strategy is described in
Note [CUDA IPC and the caching allocator].

This also fixes an unrelated bug, where we weren't actually using
the Tensor forking pickler, because we didn't register a pickler
for torch.Tensor.

Fixes #9447.  Fixes #46.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CC @apaszke

### CODE REVIEW

File: torch/serialization.py [hunk 1]
```diff
@@ -449,7 +449,18 @@ def persistent_load(saved_id):
                 storage_views = pickle_module.load(f)
                 for target_cdata, root_cdata, offset, size in storage_views:
                     root = deserialized_objects[root_cdata]
-                    deserialized_objects[target_cdata] = root[offset:offset + size]
+                    if offset != 0 or size != root.size():
+                        warnings.warn("Detected storage view in legacy serialized data: "
+                                      "storages are no longer natively supported, so we are making "
```
Comment (apaszke) [inline torch/serialization.py]: storages are supported, views aren't

File: torch/serialization.py [hunk 2]
```diff
@@ -449,7 +449,18 @@ def persistent_load(saved_id):
...
+                                      "a copy of the data instead.  THIS IS A SEMANTIC CHANGE! "
+                                      "If you need aliasing, reserialize your model using "
+                                      "tensors that share storage.")
+
+                        tensor = torch._utils._rebuild_tensor(root, offset, (size,), (1,))
+                        obj = tensor.clone().storage()
+                    else:
+                        obj = root[offset:offset + size]
```
Comment (apaszke) [inline torch/serialization.py]: Won't this always fail?

**Review by apaszke — APPROVED** [review]

File: torch/csrc/generic/StorageSharing.cpp [hunk 3]
```diff
@@ -242,41 +231,35 @@ static PyObject * THPStorage_(shareCuda)(THPStorage *self)
...
   PyObject *_size = PyTuple_GET_ITEM(args, 2);
-  PyObject *_offset = PyTuple_GET_ITEM(args, 3);
-  PyObject *_view_size = PyTuple_GET_ITEM(args, 4);
   if (!(THPUtils_checkLong(_device) && THPUtils_checkLong(_size)
-      && (_handle == Py_None || PyBytes_Check(_handle))
-      && THPUtils_checkLong(_offset) && THPUtils_checkLong(_view_size))) {
+      && (_handle == Py_None || PyBytes_Check(_handle)))) {
     THPUtils_invalidArguments(args, NULL, "_new_shared in CUDA mode", 1,
-        "(int device, bytes handle, int storage_size, int offset, int view_size");
+        "(int device, bytes handle, int storage_size)");
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/generic/StorageSharing.cpp]: You misread the diff: `shareCuda` still has offset, but `newSharedCuda` does not. This is intentional: `shareCuda` will return the offset from the base of the cuda caching allocator allocation.

File: torch/csrc/generic/StorageSharing.cpp [hunk 4]
```diff
@@ -296,11 +279,7 @@ static PyObject * THPStorage_(newSharedCuda)(PyObject *_unused, PyObject *args)
       LIBRARY_STATE
       THCIpcDeleter::makeDataPtr(devPtr, device),
       storage_size, /* allocator */ nullptr));
-  base->flag = TH_STORAGE_REFCOUNTED;
-
-  if (offset != 0 || view_size != storage_size) {
-    return THPStorage_(newTHView)(base.get(), offset, view_size);
-  }
+  base->flag = TH_STORAGE_REFCOUNTED;  // NB: Not resizable
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/generic/StorageSharing.cpp]: Oh I thought we have an "unresizable" flag, while in reality we only have a "resizable" one. Never mind my comment

**Pull request closed without merging**